### ISSUE TITLE
UNKNOWN ERROR (4) on GPU

### ISSUE THREAD

**Issue by jinluyang** [pytorch/xla#3055]
## ❓ Questions and Help
Hi, I am running on GPU using direct session. In some net works, the cuda driver throws UNKNOWN ERROR (4). I found this has something to do with ReleaseHandle and ReleaseHandles in xrt_computation_client.cc，it occurs after calling them.
My question is, there is a function TriggeredTask::Stop, but I cannot find anywhere in the code that calls it. It is never executed, why?

second, in my case, XrtComputationClient::PrepareToExit calls size_t run_id = triggered_task_->Activate(), but the triggerd_task just did nothing, because the released_data_handles_ at this time is empty. I don't know if this is normal or not, or maybe my colleague has modified some code I don't know.

third, function ReleaseHandle is called in some destruction functions, after PrepareToExit finishes. ReleaseHandle is called many times, and scheduled in the triggered_tasks_. Usually in one of the threads, It throws an error. the backtrace is like below(the linenum maybe different because I am using a customed TF):
#0  0x00007ffff78325f7 in raise () from /lib64/libc.so.6
#1  0x00007ffff7833ce8 in abort () from /lib64/libc.so.6
#2  0x00007fff4981bc44 in tensorflow::internal::LogMessageFatal::~LogMessageFatal (this=<optimized out>, __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at tensorflow/core/platform/default/logging.cc:300
#3  0x00007fff493367d9 in stream_executor::gpu::ScopedActivateContext::ScopedActivateContext (this=<optimized out>, cuda_context=<optimized out>) at tensorflow/stream_executor/cuda/cuda_driver.cc:234
#4  0x00007fff4933ab3f in stream_executor::gpu::GpuDriver::CreateStream (context=0x7fff00a44a80, stream=stream@entry=0x7ffa940112e0, priority=priority@entry=0) at tensorflow/stream_executor/cuda/cuda_driver.cc:814
#5  0x00007fff49332fc3 in stream_executor::gpu::GpuStream::Init (this=0x7ffa940112d0) at tensorflow/stream_executor/gpu/gpu_stream.cc:26
#6  0x00007fff493a7e1b in stream_executor::StreamExecutor::AllocateStream (this=0x7fff003d7690, stream=stream@entry=0x7ffa94015e30) at tensorflow/stream_executor/stream_executor_pimpl.cc:739
#7  0x00007fff49396b09 in stream_executor::Stream::Init (this=this@entry=0x7ffa94015e30) at tensorflow/stream_executor/stream.cc:298
#8  0x00007fff476b5e28 in xla::StreamPool::BorrowStream (this=0x7ffcd61c7e30, executor=executor@entry=0x7fff003d7690) at tensorflow/compiler/xla/service/stream_pool.cc:45
#9  0x00007fff46cabcf5 in xla::Backend::BorrowStream (this=this@entry=0x7ffcd62284b0, executor=<optimized out>) at tensorflow/compiler/xla/service/backend.cc:118
#10 0x00007fff46cac004 in xla::Backend::BorrowStream (this=0x7ffcd62284b0, device_ordinal=<optimized out>) at tensorflow/compiler/xla/service/backend.cc:110
#11 0x00007fff44747738 in tensorflow::XlaDevice::EnsureStreamOkLocked (this=this@entry=0x7ffa940108f0, backend=backend@entry=0x7ffcd62284b0, name=..., stream=stream@entry=0x7ffa94010ac8,
    stream_was_changed=stream_was_changed@entry=0x7fff38f2c6af) at tensorflow/compiler/jit/xla_device.cc:276
#12 0x00007fff44747ca4 in tensorflow::XlaDevice::GetDeviceContextLocked (this=this@entry=0x7ffa940108f0) at tensorflow/compiler/jit/xla_device.cc:297
#13 0x00007fff44748897 in tensorflow::XlaDevice::UseGpuDeviceInfo (this=this@entry=0x7ffa940108f0) at tensorflow/compiler/jit/xla_device.cc:361
#14 0x00007fff44732944 in tensorflow::XlaGpuDeviceFactory::CreateDevices (this=<optimized out>, session_options=..., name_prefix=..., devices=0x7fff38f2cf30) at tensorflow/compiler/jit/xla_gpu_device.cc:134
#15 0x00007fff49152b9d in tensorflow::DeviceFactory::AddDevices (options=..., name_prefix=..., devices=devices@entry=0x7fff38f2cf30) at tensorflow/core/common_runtime/device_factory.cc:139
#16 0x00007fff444d2d33 in tensorflow::DirectSessionFactory::NewSession (this=<optimized out>, options=..., out_session=0x7fff38f2d368) at tensorflow/core/common_runtime/direct_session.cc:190
#17 0x00007fff491d24f0 in tensorflow::NewSession (options=..., out_session=out_session@entry=0x7fff38f2d368) at tensorflow/core/common_runtime/session.cc:97
#18 0x00007fff412fb772 in tensorflow::ClientSession::ClientSession (this=0x7ffa94000cb0, scope=..., session_options=...) at tensorflow/cc/client/client_session.cc:56
#19 0x00007fff412d63f1 in xla::XrtSession::XrtSession (this=0x7ffa94000ca0, session_options=...) at tensorflow/compiler/xla/xla_client/xrt_session.cc:10
#20 0x00007fff412d7d0d in construct<xla::XrtSession, tensorflow::SessionOptions&> (this=<optimized out>, __p=<optimized out>)
    at /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/ext/new_allocator.h:136
#21 construct<xla::XrtSession, tensorflow::SessionOptions&> (__a=..., __p=<optimized out>) at /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/bits/alloc_traits.h:475
#22 _Sp_counted_ptr_inplace<tensorflow::SessionOptions&> (__a=..., this=0x7ffa94000c90) at /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/bits/shared_ptr_base.h:526
#23 __shared_count<xla::XrtSession, std::allocator<xla::XrtSession>, tensorflow::SessionOptions&> (__a=..., this=0x7fff38f2d8d8)
    at /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/bits/shared_ptr_base.h:637
#24 __shared_ptr<std::allocator<xla::XrtSession>, tensorflow::SessionOptions&> (__a=..., __tag=..., this=0x7fff38f2d8d0)
    at /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/bits/shared_ptr_base.h:1295
#25 shared_ptr<std::allocator<xla::XrtSession>, tensorflow::SessionOptions&> (__a=..., __tag=..., this=0x7fff38f2d8d0) at /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/bits/shared_ptr.h:344
#26 allocate_shared<xla::XrtSession, std::allocator<xla::XrtSession>, tensorflow::SessionOptions&> (__a=...) at /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/bits/shared_ptr.h:691
#27 make_shared<xla::XrtSession, tensorflow::SessionOptions&> () at /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/bits/shared_ptr.h:707
#28 xla::XrtSessionCache::CreateSession (this=this@entry=0x555559694040, target=...) at tensorflow/compiler/xla/xla_client/xrt_session_cache.cc:65
#29 0x00007fff412d8390 in xla::XrtSessionCache::GetSession (this=this@entry=0x555559694040, target=...) at tensorflow/compiler/xla/xla_client/xrt_session_cache.cc:24
#30 0x00007fff412d852f in xla::XrtSessionCache::GetSession (this=this@entry=0x555559694040, target=..., session_map=session_map@entry=0x7fff38f2db80) at tensorflow/compiler/xla/xla_client/xrt_session_cache.cc:31
#31 0x00007fff412a96a3 in xla::XrtComputationClient::GetSessionForTarget (this=this@entry=0x555559693630, cache=cache@entry=0x555559694040, target=..., session_map=session_map@entry=0x7fff38f2db80)
    at tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:1024
#32 0x00007fff412b1829 in xla::XrtComputationClient::GetSessionForXrtDevice (this=this@entry=0x555559693630, cache=0x555559694040, xrt_device=..., session_map=session_map@entry=0x7fff38f2db80)
    at tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:1031
#33 0x00007fff412b18e9 in xla::XrtComputationClient::GetSessionForDevice (this=this@entry=0x555559693630, cache=<optimized out>, device=..., session_map=session_map@entry=0x7fff38f2db80)
#34 0x00007fff412c1bf5 in xla::XrtComputationClient::ReleaseHandles(std::vector<xla::XrtComputationClient::DeviceHandle, std::allocator<xla::XrtComputationClient::DeviceHandle> >*, std::function<xla::XrtSession::CachedNode const& (xla::X
rtSession*, tensorflow::Scope const&, std::string const&)> const&, xla::metrics::Metric*, xla::metrics::Counter*) (this=this@entry=0x555559693630, handles=handles@entry=0x5555596937d0, op_generator=..., timed_metric=<optimized out>,
    destroy_counter=destroy_counter@entry=0x7ffa90000900) at tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:1192
#35 0x00007fff412c28f2 in xla::XrtComputationClient::HandleReleaser (this=0x555559693630) at tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:1238
#36 0x00007fff412a0244 in operator() (this=0x555559694460) at /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/bits/std_function.h:706
#37 xla::util::TriggeredTask::Runner (this=0x555559694460) at tensorflow/compiler/xla/xla_client/triggered_task.cc:67
#38 0x00007fff49d55b8f in execute_native_thread_routine () from /usr/local/anaconda3/lib/python3.7/site-packages/torch_xla-1.8-py3.7-linux-x86_64.egg/torch_xla/lib/libxla_computation_client.so
#39 0x00007ffff7bc6dc5 in start_thread () from /lib64/libpthread.so.0
#40 0x00007ffff78f3ced in clone () from /lib64/libc.so.6

it cannot create ClientSession. I think I'll try skipping those ReleaseHandles. Some helps would be appreciated, thanks.

**Comment by jinluyang** [comment]
And by the way, is there any conclusion on which part of building the graph of ir::Node consumes most of the time? Would like to try some optimization to accelerate inference.

**Comment by jinluyang** [comment]
Hi, I have quited on the issue of ReleaseHandles, since I found the overhead of building the graph is very big compared to the GPU kernel. Maybe the best way of optimizing inference is to store the graph in some data structure, which is no longer a lazy-tensor approach.

**Comment by JackCaoG** [comment]
Hmm, graph building should be overlapped with the kernel execution. If you have more info to share(what percentage of time that graph building has taken etc) we'd love to know.

**Comment by jinluyang** [comment]
> Hmm, graph building should be overlapped with the kernel execution. If you have more info to share(what percentage of time that graph building has taken etc) we'd love to know.

In a network which consists of a loop of tril、matmul and mult，the graph building time each loop is about 560 us, and the kernel time is about 81 us. I have a .qdrep file with nvtx. 
I also tried a part of a gpt model, the graph-building time is 115ms, and the estimated kernel time (because there are some bugs, I couldn't directly run it, I used tensorflow XLA instead) is 5.99ms.
I also tried to build a light-weight graph. Instead of going deeper to the GetIrValue functions, I want to record the operator message in the functions in aten_xla_type.cpp, and when the output is required, compare the light-weight graph to a cache and decide whether to build the graph and compile, or run executable directly. I tried to directly return a dummy aten::Tensor in the functions in aten_xla_type.cpp. the time decreased to 26.27ms, but still big compared to 5.99ms.
